orchestrator/graphql/resolvers/process.py

```diff
@@ -0,0 +1,224 @@
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module that implements process related API endpoints."""
+
+from enum import Enum
+from http import HTTPStatus
+from uuid import UUID
+
+import strawberry
+import structlog
+from more_itertools import chunked, flatten
+from sqlalchemy import String, cast
+from sqlalchemy.orm import defer, joinedload
+from sqlalchemy.sql import expression
+
+from orchestrator.api.error_handling import raise_status
+from orchestrator.api.helpers import VALID_SORT_KEYS
+from orchestrator.db import ProcessSubscriptionTable, ProcessTable, ProductTable, SubscriptionTable, db
+from orchestrator.db.database import SearchQuery
+from orchestrator.graphql.pagination import Connection, PageInfo
+from orchestrator.graphql.schemas.process import ProcessType
+from orchestrator.schemas import ProcessSchema
+from orchestrator.schemas.process import ProcessStepSchema
+
+logger = structlog.get_logger(__name__)
+
+
+def enrich_process(p: ProcessTable) -> ProcessSchema:
+    # p.subscriptions is a non JSON serializable AssociationProxy
+    # So we need to build a list of Subscriptions here.
+    return ProcessSchema(
+        assignee=p.assignee,
+        created_by=p.created_by,
+        failed_reason=p.failed_reason,
+        last_modified=p.last_modified_at,
+        id=p.pid,
+        started=p.started_at.timestamp(),
+        last_step=p.last_step,
+        workflow_name=p.workflow,
+        is_task=p.is_task,
+        status=p.last_status,
+        steps=[
+            ProcessStepSchema(
+                stepid=step.stepid,
+                name=step.name,
+                status=step.status,
+                created_by=step.created_by,
+                executed=step.executed_at,
+                commit_hash=step.commit_hash,
+                state=step.state,
+            )
+            for step in p.steps
+        ],
+        current_state={},
+        subscriptions=[],
+    )
+
+
+@strawberry.enum(description="Sort order (ASC or DESC)")
+class SortOrder(Enum):
+    ASC = "asc"
+    DESC = "desc"
+
+
+@strawberry.input(description="Sort processes by attribute")
+class ProcessSort:
+    field: str = strawberry.field(description="Field to sort on")
+    order: SortOrder = strawberry.field(default=SortOrder.ASC, description="Sort order (ASC or DESC")
+
+
+def filter_processes(query: SearchQuery, filter: list[str] | None = None) -> SearchQuery:
+    if filter is not None:
+        if len(filter) == 0 or (len(filter) % 2) > 0:
+            raise_status(HTTPStatus.BAD_REQUEST, "Invalid number of filter arguments")
+
+        for filter_pair in chunked(filter, 2):
```

Comment:
I didn't even start to read this function since the for loop is way to big.

Reply:
I have refactored it and also added the new way of filtering to the rest endpoint